trl/trainer/grpo_config.py

```diff
@@ -112,9 +112,19 @@ class GRPOConfig(TrainingArguments):
         scale_rewards (`bool`, *optional*, defaults to `True`):
             Whether to scale the rewards by dividing them by their standard deviation. If `True` (default), the rewards
             are normalized by the standard deviation, ensuring they have unit variance. If `False`, no scaling is
-            applied. The [Dr. GRPO](https://github.com/sail-sg/understand-r1-zero/blob/main/understand-r1-zero.pdf)
-            paper recommends not scaling the rewards, as scaling by the standard deviation introduces a question-level
-            difficulty bias.
+            applied. The [Dr. GRPO paper](https://huggingface.co/papers/2503.14476) recommends not scaling the rewards,
+            as scaling by the standard deviation introduces a question-level difficulty bias.
+        loss_type (`str`, *optional*, defaults to `"bnpo"`):
```

Comment:
What is `bnpo`? Would be good to have a reference to where it's defined (I thought we had DAPO as the default loss)

Reply:
In fact, I realized while doing this PR that it wasn't exactly DAPO that was being used, but a variant of BNPO as defined here :

<img width="1063" alt="Screenshot 2025-04-08 at 06 27 50" src="https://github.com/user-attachments/assets/e52d1dd8-80ae-4c23-8006-b1c606c1fdf0" />

Let me try to clarify here. Losses per token are normalized by 
- GRPO: the length of the sequence
- DAPO: the average sequence length in the group
- BNPO: the average sequence length in the batch
- TRL: the average sequence length in the local batch*; this is what I call bnpo here, but it's not 100% correct
- Dr GRPO: by the maximum possible length of the completion 
- 
- * a batch is made up of `num_devices * gradient_accumulations` local batches

Reply:
Special cases:
When
- `per_device_batch_size==num_generations`, TRL's BNPO is equivalent to DAPO
- `per_device_batch_size==1`, TRL's BNPO is equivalent to GRPO
- `gradient_accumualtion_steps==1` and `num_devices=1`,  TRL's BNPO is equivalent to the actual BNPO.

Reply:
@qgallouedec  Thanks for the comprehensive support! A minor comment for your future consideration: Dr. GRPO does not constrain the constant normalizer to be MAX_LEN (although it's easier to just use that). This can affect the update scale (related to your recent tweet https://x.com/QGallouedec/status/1908741708021457357). In fact, different constant of `x` in the setting in your tweet can be absorbed into the constant normalizer we propose in the paper, and MAX_LEN is a convenient example.